### DISCUSSION TITLE
MVVM ViewModels in separate sloution project

### DISCUSSION THREAD

**Discussion by SoftEntwickler** [microsoft/TemplateStudio#4619] [General]
Hi,
Would it not make sense to move the ViewModels in a separate solution project?

**Comment by mrlacey** [comment]
"make sense" is subjective.

I've known projects that do and don't do this. Mostly it comes down to developer preference.

Generally, it's useful to have ViewModels in a separate project when there's a specific benefit to doing so. The benefits I've seen of putting the VMs in a separate project relate to the separation and isolation of the VM from platform-specific functionality.
This is useful when the VMs are going to be shared in multiple projects that target different platforms. This is not a primary goal of Template Studio.
Having VMs without any platform-specific knowledge/functionality/types can also lead to more complicated code, as conversions are required.

In that Template Studio tries to be useful to both new and experienced developers, adding more projects to the generated solution potentially makes things more complicated 
Additionally, in that Template Studio tries to demonstrate best practices for producing Windows Apps in its output. I don't think that making such a change would do that.

You are, of course, free to move the files from the generated solution in whatever way you wish. IMO. In terms of changing what Template Studio produces, it would likely be a lot of work and for limited (if any) tangible benefits.

Are there other benefits of making this change that I'm not aware of?

**Reply by SoftEntwickler** [reply]
Thank you for your replay.
I'm only developing windows desktop applications and even there I always create a separate project for ViewModels, because it helps me to be more disciplined during design my code and take care for concerns in a cleaner way. So, the additional benefit is just handling and dealing with a bigger codebase. For heavy system depended ViewModels I think it is ok to create them in the application project, but in ViewModels namespace.

e.g. 
- ProjectA
-- Folder: ViewModels
-- Folder: Views
- ProjectA.ViewModels

It took me a long time to figure out how to structure the solution projects. I think for new people it would be helpful to see how they can separate their code in different projects.

**Reply by mikebattista** [reply]
To add on to @mrlacey's reply, adding more options increases the permutations that need to be managed in the templates, so we need to be thoughtful when adding options and avoiding excessive cost and maintenance unless it adds a lot of value.

In this case, you can easily move the ViewModels to a separate project if you wish, and Template Studio can still help generate the majority of the code for you when you add a New Page to the project via the Add -> New Item (Template Studio) -> New Page wizard.
